SmartDeviceLink/SDLTCPTransport.m

```diff
@@ -52,125 +53,303 @@ - (instancetype)initWithHostName:(NSString *)hostName portNumber:(NSString *)por
-        SDLTCPTransport *transport = (__bridge SDLTCPTransport *)info;
-        [transport.delegate onTransportConnected];
-    } else if (kCFSocketDataCallBack == type) {
-        SDLTCPTransport *transport = (__bridge SDLTCPTransport *)info;
+- (void)sdl_onConnectionTimedOut:(NSTimer *)timer {
+    SDLLogW(@"TCP connection timed out");
+    [self sdl_cancelIOThread];
 
-        // Check if Core disconnected from us
-        if (CFDataGetLength((CFDataRef)data) <= 0) {
-            SDLLogW(@"Remote system terminated connection, data packet length 0");
-            [transport.delegate onTransportDisconnected];
+    if (!self.transportErrorNotified) {
+        if (!self.transportConnected) { // this condition should be always true
+            SDLLogD(@"Notifying connection timed out error");
+            [self.delegate onError:[NSError sdl_transport_connectionTimedOutError]];
+        } else {
+            [self.delegate onTransportDisconnected];
+        }
+        self.transportErrorNotified = YES;
+    }
+}
 
-            return;
+- (void)sdl_onStreamError:(NSStream *)stream {
+    // stop I/O thread
+    [self sdl_cancelIOThread];
+
+    // avoid notifying multiple error events
+    if (!self.transportErrorNotified) {
+        if (self.transportConnected) {
+            // transport is disconnected while running
+            [self.delegate onTransportDisconnected];
+            self.transportErrorNotified = YES;
+        } else if ([stream.streamError.domain isEqualToString:NSPOSIXErrorDomain]) {
+            // connection error
+
+            // According to Apple's document "Error Objects, Domains, and Codes", the 'code' values
+            // of NSPOSIXErrorDomain are actually errno values.
+            NSError *error;
+            switch (stream.streamError.code) {
+                case ECONNREFUSED:
+                    SDLLogD(@"TCP connection error: ECONNREFUSED");
+                    error = [NSError sdl_transport_connectionRefusedError];
+                    break;
+                case ETIMEDOUT:
+                    SDLLogD(@"TCP connection error: ETIMEDOUT");
+                    error = [NSError sdl_transport_connectionTimedOutError];
+                    break;
+                case ENETDOWN:
+                    SDLLogD(@"TCP connection error: ENETDOWN");
+                    error = [NSError sdl_transport_networkDownError];
+                    break;
+                case ENETUNREACH:
+                    // This is just for safe. I did not observe ENETUNREACH error on iPhone.
+                    SDLLogD(@"TCP connection error: ENETUNREACH");
+                    error = [NSError sdl_transport_networkDownError];
+                    break;
+                default:
+                    SDLLogD(@"TCP connection error: unknown error %ld", (long)stream.streamError.code);
+                    error = [NSError sdl_transport_OthersError];
+                    break;
+            }
+            [self.delegate onError:error];
```

Comment:
Question: All the errors here are errors causing us to not be able to connect, should we be checking anything here to make sure we're disconnected?

Reply:
Sorry, I didn't catch your question. Are you suggesting that we should verify `stream.streamStatus` property here and make sure that the stream is closed? 

Well, once `sdl_onStreamError:` is triggered, the `ioThread` is cancelled and eventually `sdl_teardownStream:` will be called for both streams. So I think there may not be much point adding extra logic inside `sdl_onStreamError:`.